File: mifosng-android/src/main/java/com/mifos/mifosxdroid/offline/offlinedashbarod/OfflineDashboardPresenter.java
package com.mifos.mifosxdroid.offline.offlinedashbarod;

import com.mifos.api.datamanager.DataManagerCenter;
import com.mifos.api.datamanager.DataManagerClient;
import com.mifos.api.datamanager.DataManagerGroups;
import com.mifos.api.datamanager.DataManagerLoan;
import com.mifos.api.datamanager.DataManagerSavings;
import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.accounts.loan.LoanRepaymentRequest;
import com.mifos.objects.accounts.savings.SavingsAccountTransactionRequest;
import com.mifos.objects.client.ClientPayload;
import com.mifos.objects.group.GroupPayload;
import com.mifos.services.data.CenterPayload;

import java.util.List;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by Rajan Maurya on 20/07/16.
 */
public class OfflineDashboardPresenter extends BasePresenter<OfflineDashboardMvpView> {


    private CompositeSubscription mSubscriptions;
    private final DataManagerClient mDataManagerClient;
    private final DataManagerGroups mDataManagerGroups;
    private final DataManagerCenter mDataManagerCenters;
    private final DataManagerLoan mDataManagerLoan;
    private final DataManagerSavings mDataManagerSavings;

    @Inject
    public OfflineDashboardPresenter(DataManagerClient dataManagerClient,
                                     DataManagerGroups dataManagerGroups,
                                     DataManagerCenter dataManagerCenter,
                                     DataManagerLoan dataManagerLoan,
                                     DataManagerSavings dataManagerSavings) {
        mSubscriptions = new CompositeSubscription();
        mDataManagerClient = dataManagerClient;
        mDataManagerGroups = dataManagerGroups;
        mDataManagerCenters = dataManagerCenter;
        mDataManagerLoan = dataManagerLoan;
        mDataManagerSavings = dataManagerSavings;
    }

    @Override
    public void attachView(OfflineDashboardMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        mSubscriptions.unsubscribe();
    }

    public void loadDatabaseClientPayload() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscriptions.add(mDataManagerClient.getAllDatabaseClientPayload()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<ClientPayload>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(R.string.failed_to_load_clientpayload);
                    }

                    @Override
                    public void onNext(List<ClientPayload> clientPayloads) {
                        getMvpView().showClients(clientPayloads);
                        getMvpView().showProgressbar(false);
                    }
                }));
    }

    public void loadDatabaseGroupPayload() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscriptions.add(mDataManagerGroups.getAllDatabaseGroupPayload()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<GroupPayload>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(R.string.failed_to_load_grouppayload);
                    }

                    @Override
                    public void onNext(List<GroupPayload> groupPayloads) {
                        getMvpView().showGroups(groupPayloads);
                        getMvpView().showProgressbar(false);
                    }
                }));
    }

    public void loadDatabaseCenterPayload() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscriptions.add(mDataManagerCenters.getAllDatabaseCenterPayload()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<CenterPayload>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(R.string.failed_to_load_centerpayload);
                    }

                    @Override
                    public void onNext(List<CenterPayload> centerPayloads) {
                        getMvpView().showCenters(centerPayloads);
                        getMvpView().showProgressbar(false);
                    }
                }));
    }

    public void loadDatabaseLoanRepaymentTransactions() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscriptions.add(mDataManagerLoan.getDatabaseLoanRepayments()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<LoanRepaymentRequest>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(R.string.failed_to_load_loanrepayment);
                    }

                    @Override
                    public void onNext(List<LoanRepaymentRequest> loanRepaymentRequests) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showLoanRepaymentTransactions(loanRepaymentRequests);
                    }
                }));
    }

    public void loadDatabaseSavingsAccountTransactions() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        mSubscriptions.add(mDataManagerSavings.getAllSavingsAccountTransactions()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<SavingsAccountTransactionRequest>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(R.string.failed_to_load_savingaccounttransaction);
                    }

                    @Override
                    public void onNext(List<SavingsAccountTransactionRequest>
                                               transactionRequests) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showSavingsAccountTransaction(transactionRequests);
                    }
                })
        );
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/offline/offlinedashbarod/OfflineDashboardPresenter.kt
package com.mifos.mifosxdroid.offline.offlinedashbarod

import com.mifos.api.datamanager.DataManagerCenter
import com.mifos.api.datamanager.DataManagerClient
import com.mifos.api.datamanager.DataManagerGroups
import com.mifos.api.datamanager.DataManagerLoan
import com.mifos.api.datamanager.DataManagerSavings
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.accounts.loan.LoanRepaymentRequest
import com.mifos.objects.accounts.savings.SavingsAccountTransactionRequest
import com.mifos.objects.client.ClientPayload
import com.mifos.objects.group.GroupPayload
import com.mifos.services.data.CenterPayload
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 20/07/16.
 */
class OfflineDashboardPresenter @Inject constructor(
    dataManagerClient: DataManagerClient,
    dataManagerGroups: DataManagerGroups,
    dataManagerCenter: DataManagerCenter,
    dataManagerLoan: DataManagerLoan,
    dataManagerSavings: DataManagerSavings
) : BasePresenter<OfflineDashboardMvpView>() {
    private val mSubscriptions: CompositeSubscription = CompositeSubscription()
    private val mDataManagerClient: DataManagerClient
    private val mDataManagerGroups: DataManagerGroups
    private val mDataManagerCenters: DataManagerCenter
    private val mDataManagerLoan: DataManagerLoan
    private val mDataManagerSavings: DataManagerSavings

    init {
        mDataManagerClient = dataManagerClient
        mDataManagerGroups = dataManagerGroups
        mDataManagerCenters = dataManagerCenter
        mDataManagerLoan = dataManagerLoan
        mDataManagerSavings = dataManagerSavings
    }

    override fun attachView(mvpView: OfflineDashboardMvpView) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        mSubscriptions.unsubscribe()
    }

    fun loadDatabaseClientPayload() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(mDataManagerClient.allDatabaseClientPayload
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<ClientPayload>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    mvpView!!.showProgressbar(false)
                    mvpView!!.showError(R.string.failed_to_load_clientpayload)
                }

                override fun onNext(clientPayloads: List<ClientPayload>) {
                    mvpView!!.showClients(clientPayloads)
                    mvpView!!.showProgressbar(false)
                }
            })
        )
    }

    fun loadDatabaseGroupPayload() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(mDataManagerGroups.allDatabaseGroupPayload
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<GroupPayload>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    mvpView!!.showProgressbar(false)
                    mvpView!!.showError(R.string.failed_to_load_grouppayload)
                }

                override fun onNext(groupPayloads: List<GroupPayload>) {
                    mvpView!!.showGroups(groupPayloads)
                    mvpView!!.showProgressbar(false)
                }
            })
        )
    }

    fun loadDatabaseCenterPayload() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(mDataManagerCenters.allDatabaseCenterPayload
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<CenterPayload>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    mvpView!!.showProgressbar(false)
                    mvpView!!.showError(R.string.failed_to_load_centerpayload)
                }

                override fun onNext(centerPayloads: List<CenterPayload>) {
                    mvpView!!.showCenters(centerPayloads)
                    mvpView!!.showProgressbar(false)
                }
            })
        )
    }

    fun loadDatabaseLoanRepaymentTransactions() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(mDataManagerLoan.databaseLoanRepayments
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<LoanRepaymentRequest>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    mvpView!!.showProgressbar(false)
                    mvpView!!.showError(R.string.failed_to_load_loanrepayment)
                }

                override fun onNext(loanRepaymentRequests: List<LoanRepaymentRequest>) {
                    mvpView!!.showProgressbar(false)
                    mvpView!!.showLoanRepaymentTransactions(loanRepaymentRequests)
                }
            })
        )
    }

    fun loadDatabaseSavingsAccountTransactions() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(
            mDataManagerSavings.allSavingsAccountTransactions
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<SavingsAccountTransactionRequest>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showError(R.string.failed_to_load_savingaccounttransaction)
                    }

                    override fun onNext(transactionRequests: List<SavingsAccountTransactionRequest>) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showSavingsAccountTransaction(transactionRequests)
                    }
                })
        )
    }
}